utils/dataset.py

```diff
@@ -171,23 +207,105 @@ def __len__(self) -> int:
+        text = _stack_to_tensor(text).long()
+        spec_lens = [b['mel_len'] for b in batch]
+        max_spec_len = max(spec_lens) + 1
+        if max_spec_len % self.r != 0:
+            max_spec_len += self.r - max_spec_len % self.r
+        mel = [_pad2d(b['mel'], max_spec_len) for b in batch]
+        mel = _stack_to_tensor(mel)
+        item_id = [b['item_id'] for b in batch]
+        speaker_name = [b['speaker_name'] for b in batch]
+        mel_lens = [b['mel_len'] for b in batch]
+        mel_lens = torch.tensor(mel_lens)
+        speaker_emb = [b['speaker_emb'] for b in batch]
+        speaker_emb = _stack_to_tensor(speaker_emb)
+
+        return {'x': text, 'mel': mel, 'item_id': item_id,
+                'x_len': x_len, 'mel_len': mel_lens,
+                'speaker_emb': speaker_emb, 'speaker_name': speaker_name}
+
+
+class ForwardCollator:
+
+    def __init__(self, taco_collator: TacoCollator) -> None:
+        self.taco_collator = taco_collator
+
+    def __call__(self, batch: List[Dict[str, Union[str, torch.tensor]]]) -> Dict[str, torch.tensor]:
+        output = self.taco_collator(batch)
+        x_len = [b['x_len'] for b in batch]
+        x_len = torch.tensor(x_len)
+        max_x_len = max(x_len)
+        dur = [_pad1d(b['dur'][:max_x_len], max_x_len) for b in batch]
+        dur = _stack_to_tensor(dur).float()
+        pitch = [_pad1d(b['pitch'][:max_x_len], max_x_len) for b in batch]
+        pitch = _stack_to_tensor(pitch).float()
+        energy = [_pad1d(b['energy'][:max_x_len], max_x_len) for b in batch]
+        energy = _stack_to_tensor(energy).float()
+        pitch_cond = [_pad1d(b['pitch_cond'][:max_x_len], max_x_len) for b in batch]
+        pitch_cond = _stack_to_tensor(pitch_cond).long()
+        output.update({
+            'pitch': pitch,
+            'energy': energy,
+            'dur': dur,
+            'pitch_cond': pitch_cond
+        })
+        return output
+
+
+def get_taco_dataloaders(paths: Paths,
+                         batch_size: int,
+                         r: int,
+                         max_mel_len: int,
+                         filter_duration_stats: bool,
+                         min_attention_alignment: float,
+                         min_attention_sharpness: float,
+                         max_consecutive_ones: int,
+                         max_duration: int,
+                         num_workers=0) -> Tuple[DataLoader, DataLoader]:
+
+    """Returns training and validation dataloaders.
+
+    Args:
+        paths (Paths): An instance of the Paths class containing file paths.
+        batch_size (int): The batch size for the dataloaders.
+        max_mel_len (int, optional): The maximum length of the mel spectrograms. Defaults to 1250.
+        filter_duration_stats (bool): Whether to filter accordint to below duration stats.
+        min_attention_alignment (float): The minimum attention alignment value. Defaults to 0.5.
```

Comment:
I think more organic would be to set those default values in the function definition itself

Reply:
Removed the defaults from docstring. There is actually no default, so values should be taken from the config.